### DISCUSSION TITLE
Main actor-isolated static property 'liveValue' cannot be used to satisfy nonisolated protocol requirement; this is an error in the Swift 6 language mode

### DISCUSSION THREAD

**Discussion by bitomule** [pointfreeco/swift-dependencies#310] [Q&A] [answered]
Hello!

I have started a new app and I'm facing some issues with Dependencies and a dependency marked with MainActor. I have a CoreData client marked as MainActor and I'm trying to use it in a dependency like this:

````swift
protocol ObserveDuplicateGroupsUseCase {
    func execute() async -> AsyncThrowingStream<[[String]], Error>
}

@MainActor
struct ObserveDuplicateGroupsUseCaseLive: ObserveDuplicateGroupsUseCase, Sendable {
    let client: CoreDataClient
    
    func execute() async -> AsyncThrowingStream<[[String]], Error> {
        // Code fetching data here
    }
}

private enum ObserveDuplicateGroupsUseCaseKey: DependencyKey {
    @MainActor
    static let liveValue: any ObserveDuplicateGroupsUseCase = ObserveDuplicateGroupsUseCaseLive(
        client: CoreDataClient.buildDefault()
    )
}
````

This logs a warning "Main actor-isolated static property 'liveValue' cannot be used to satisfy nonisolated protocol requirement; this is an error in the Swift 6 language mode" and removing the MainActor from liveValue transform warning into this error: "Call to main actor-isolated static method 'buildDefault()' in a synchronous nonisolated context".

What's the best way to handle a case like this?

**Comment by mbrandonw** [comment, marked as answer]
Hi @bitomule, currently `@Dependency` can be used from basically anywhere, and that means `liveValue` (and `testValue`/`previewValue`) must be non-isolated. You cannot have a dependency that is `@MainActor` bound because nothing is stopping you from doing:

```swift
// Not @MainActor
class Model { 
  @Dependency(\.mainActorClient) var …
}
```

This would require initializing a `@MainActor` thing from a non-necessarily `@MainActor` context.

Theoretically the tools in the library could start holding onto their isolation as `(any Actor)?`, but it's a big unknown of what that would do to ergonomics, and I don't think Swift's isolation tools are even up for the challenge. You would have static information of the type of isolation for a dependency, but that doesn't say anything about the _instance_ of isolation. Non-global actors can have multiple instances and each instance is a unique isolation domain.

However, typically one does not need to make the dependency itself `@MainActor`, and instead it can hold onto something that is `@MainActor` and provide `@MainActor` endpoints for accessing its state. I would suggest exploring ways of designing your dependency that does not require a `@MainActor` initializer, and then things should work just fine.

**Reply by bitomule** [reply]
Thanks! I'll explore how can I change this init

**Reply by bitomule** [reply]
If anyone is interested or facing a similar issue I just noticed my init is safe, added nonissolated and solved, issue gone. My class state is still under MainActor and safe.

**Reply by mbrandonw** [reply]
That's a good point, thanks for sharing! If a `@MainActor` type's properties all have defaults, then you can have a `nonisolated init` and that allows you to create instances anywhere.
